xds/internal/balancer/outlierdetection/balancer.go

```diff
@@ -105,3 +135,793 @@ func (bb) ParseConfig(s json.RawMessage) (serviceconfig.LoadBalancingConfig, err
+		// 1. Remove Subchannel from Addresses map entry if present in Addresses map.
+		b.removeSubConnFromAddressesMapEntry(scw)
+		// 2. Clear the Subchannel wrapper's Call Counter entry.
+		addrInfo := (*addressInfo)(atomic.LoadPointer(&scw.addressInfo))
+		if addrInfo != nil {
+			addrInfo.callCounter.clear()
+		}
+		// 3. Uneject the Subchannel in case it was previously ejected.
+		scw.uneject()
+	case len(addrs) == 1: // multiple/no addresses to single address
+		// 1. Add Subchannel to Addresses map entry if new address present in map.
+		addrInfo := b.appendIfPresent(addrs[0].Addr, scw)
+		if addrInfo != nil && !addrInfo.latestEjectionTimestamp.IsZero() {
+			scw.eject()
+		}
+	} // otherwise multiple/no addresses to multiple/no addresses; ignore
+
+	scw.addresses = addrs
+}
+
+func (b *outlierDetectionBalancer) ResolveNow(opts resolver.ResolveNowOptions) {
+	b.cc.ResolveNow(opts)
+}
+
+func (b *outlierDetectionBalancer) Target() string {
+	return b.cc.Target()
+}
+
+func max(x, y int64) int64 {
+	if x < y {
+		return y
+	}
+	return x
+}
+
+func min(x, y int64) int64 {
+	if x < y {
+		return x
+	}
+	return y
+}
+
+// handleSubConnUpdate stores the recent state and forward the update
+// if the SubConn is not ejected.
+func (b *outlierDetectionBalancer) handleSubConnUpdate(u *scUpdate) {
+	scw := u.scw
+	scw.latestState = u.state
+	b.childMu.Lock()
+	if !scw.ejected && b.child != nil {
+		b.child.UpdateSubConnState(scw, u.state)
+	}
+	b.childMu.Unlock()
+}
+
+// handleEjectedUpdate handles any SubConns that get ejected/unejected, and
+// forwards the appropriate corresponding subConnState to the child policy.
+func (b *outlierDetectionBalancer) handleEjectedUpdate(u *ejectedUpdate) {
+	scw := u.scw
+	scw.ejected = u.ejected
+	var stateToUpdate balancer.SubConnState
+	if u.ejected {
+		// "The wrapper will report a state update with the
+		// TRANSIENT_FAILURE state, and will stop passing along
+		// updates from the underlying subchannel."
+		stateToUpdate = balancer.SubConnState{
+			ConnectivityState: connectivity.TransientFailure,
+		}
+	} else {
+		// "The wrapper will report a state update with the latest
+		// update from the underlying subchannel, and resume passing
+		// along updates from the underlying subchannel."
+		stateToUpdate = scw.latestState // If this has never been written to will send connectivity IDLE which seems fine to me
+	}
+	b.childMu.Lock()
+	if b.child != nil {
+		b.child.UpdateSubConnState(scw, stateToUpdate)
+	}
+	b.childMu.Unlock()
+}
+
+// handleChildStateUpdate forwards the picker update wrapped in a wrapped picker
+// with the noop picker bit present.
+func (b *outlierDetectionBalancer) handleChildStateUpdate(u balancer.State) {
+	b.childState = u
+	b.mu.Lock()
+	noopCfg := b.noopConfig()
+	b.mu.Unlock()
+	b.recentPickerNoop = noopCfg
+	b.cc.UpdateState(balancer.State{
+		ConnectivityState: b.childState.ConnectivityState,
+		// The outlier_detection LB policy will provide a picker that delegates to
```

Comment:
The comments in this function and the one in `handleLBConfigUpdate` are the same. The `pickerWrapper` type is already documented. If you feel you add something more there, please do. But these comments here are making it hard to read the code.

Reply:
Yeah, good point. Moved no-op config comment to wrapped picker top level comment. Deleted the comments on the Update struct.